Added support for document updates and upserts [v4.1.0]

jasonbosco opened the pull request:
## Change Summary
Resolves #2

jasonbosco commented:
@AbdullahFaqeir Still working on adding some finishing touches to v0.16 of Typesense server, so this PR is not ready to release yet. But in the meantime, could you help review these code changes when you get a chance?

AbdullahFaqeir commented:
HI @jasonbosco 

I'm sorry I didn't get time to review your php code, I'm really sorry.

AbdullahFaqeir commented on src/ApiCall.php:
```diff
@@ -105,41 +112,70 @@ public function get(string $endPoint, array $params, bool $asJson = true)
      * @return array
      * @throws TypesenseClientError|GuzzleException
      */
-    public function delete(string $endPoint): array
```
Why removing the delete method for patch instead of adding patch as a new method.

jasonbosco replied:
The diff makes it look like I removed it, but I actually just _moved_ `delete` from further down in the file to be close to the other http methods in `ApiCall`:

https://github.com/typesense/typesense-php/pull/3/files#diff-198237dfe5521226a362027889383cb3294ed90976c8de7190dcf0e15f1bb3a1R174-R178

AbdullahFaqeir commented on src/Documents.php:
```diff
@@ -62,41 +62,83 @@ private function endPointPath(string $action = ''): string
      * @throws TypesenseClientError|GuzzleException|\JsonException
      */
-    public function createMany(array $documents): array
+    public function createMany(array $documents, array $options = []): array
```
I've noticed that `createMany` method will be deprecated, we need to update the laravel driver since this will brake it.
And `import` method will work instead of `createMany`?

jasonbosco replied:
`createMany` is deprecated, but not removed in this version - it will just call `import` internally. So this version of the PHP library will not break backward compatibility.

> And import method will work instead of createMany?

Correct. If you just change `createMany` to `import` it should work with no additional changes.

AbdullahFaqeir commented on src/Documents.php:
```diff
@@ -62,41 +62,83 @@ private function endPointPath(string $action = ''): string
      */
-    public function import(string $documents): string
+    public function import($documents, array $options = [])
     {
```
$options added here, needs to be referenced in the documentation of the client.

jasonbosco replied:
Yup, it's documented here: 

https://github.com/typesense/typesense-website/blob/1aa626f37491fa41dac969bfbc34dc705405d30f/typesense.org/docs/0.16.0/api.html#L1246-L1254

AbdullahFaqeir commented on README.md:
```diff
@@ -20,6 +20,7 @@ Here are some examples that walk you through how to use the client: [doc/example
 
 | Typesense Server | typesense-php |
 |------------------|----------------|
+| \>= v0.16.0 | \>= v4.1.0 |
```
since some of the changes will break earlier implementations of the client, I guess we need to create a major update for the client with each major update of the server.

jasonbosco replied:
I've added default values for the new `options` parameter in `import` and `createMany` will still work in this version, by calling `import` internally. So this version should not break any earlier implementations. Let me know if you notice any other changes that might be backward incompatible.

jasonbosco commented:
> HI @jasonbosco
> 
> I'm sorry I didn't get time to review your php code, I'm really sorry.
> 
> Update: I just went over the updates you've made all seems good, and I just want to make sure that you ran the examples and everything went well, l guess I need to work on the laravel package before releasing the update for typesense 0.16.0 to make it compatible.

@AbdullahFaqeir No worries, thank you for reviewing the code. I've added my comments to the PR. The changes I've made here should be backward compatible with previous implementations since I've added default values to new method arguments and only added a deprecating warning log message to `createMany`. So we should be able to release the PHP library independently. But let me know if you see any other issues with backward incompatibility. The goal is to make this release backward compatible.

AbdullahFaqeir commented:
> > HI @jasonbosco
> > I'm sorry I didn't get time to review your php code, I'm really sorry.
> > Update: I just went over the updates you've made all seems good, and I just want to make sure that you ran the examples and everything went well, l guess I need to work on the laravel package before releasing the update for typesense 0.16.0 to make it compatible.
> 
> @AbdullahFaqeir No worries, thank you for reviewing the code. I've added my comments to the PR. The changes I've made here should be backward compatible with previous implementations since I've added default values to new method arguments and only added a deprecating warning log message to `createMany`. So we should be able to release the PHP library independently. But let me know if you see any other issues with backward incompatibility. The goal is to make this release backward compatible.

@jasonbosco All is good, thanks mate.

Cheers,
